File: src/api/state.rs
// Copyright (c) 2025 Proton AG
//
// This file is part of ProtonVPN.
//
// ProtonVPN is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// ProtonVPN is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.

/// Combined state of the VPN connection and the TUN interface.
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[derive(Clone, Debug, PartialEq)]
pub struct VpnState {
    pub interface_state: InterfaceState,
    pub connection_state: ConnectionState,
}

/// State of the TUN interface.
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[derive(Clone, Debug, PartialEq)]
pub struct InterfaceState {
    pub is_up: bool,
}

/// State of the VPN connection.
#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[derive(Clone, Debug, PartialEq)]
pub enum ConnectionState {

    /// Disconnected. [error] will be set if disconnection happened due to an error.
    Disconnected {
        error: Option<DisconnectReason>
    },

    /// Library is attempting VPN connection to one or more candidate peers.
    Connecting {
        peers: Vec<PeerConnectionInfo>,
        wait_reasons: Vec<PeerConnectionWaitReason>,
    },

    /// Connection to [peer] is established.
    Connected {
        peer: PeerConnectionInfo,
    },
}

#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
#[derive(PartialEq, Clone, Debug)]
pub struct PeerConnectionInfo {
    pub peer_id: String,
    pub entry_ip: String,
    pub protocol: Protocol,
    pub port: u16,
}

#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[derive(Clone, Debug, PartialEq)]
pub enum PeerConnectionWaitReason {

    /// Device currently has no network (airplane mode, no signal, etc.)
    WaitingForNetwork,

    /// There is I/O problem with TUN interface. Calling code might need to wait, recreate TUN or
    /// disconnect (when it was caused by connection by another VPN app).
    TunIoError { message: String },
}

#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[derive(Clone, Debug, PartialEq)]
pub enum DisconnectReason {

    /// There was a problem establishing TUN interface.
    TunEstablishError { message: String },
}

#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
#[derive(Debug, Copy, Clone, PartialEq)]
pub enum Protocol {
    WireguardUdp, WireguardTcp, Stealth,
}

File: sdk-android/sdk/src/main/java/me/proton/vpn/sdk/api/State.kt
/*
 * Copyright (c) 2025. Proton AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.proton.vpn.sdk.api

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.util.Date

@Parcelize
sealed interface VpnConnectionState : Parcelable {

    /**
     * VPN state is being loaded (VPN service might be running and connected but this information
     * is being collected)
     */
    data object Loading: VpnConnectionState

    data class Disconnected(
        /** != null if the disconnection was due to an error */
        val error: VpnDisconnectError? = null
    ): VpnConnectionState

    data class Connecting(
        val connections: List<PeerConnection>
    ): VpnConnectionState

    /**
     * Connection attempt requires app, user or system action to proceed.
     */
    data class WaitingForAction(
        val reason: VpnWaitReason
    ): VpnConnectionState

    data class Connected(
        val connection: PeerConnection,
        val connectedSince: Date
    ): VpnConnectionState
}

@Parcelize
sealed interface VpnWaitReason : Parcelable {

    /**
     * Device currently has no network (airplane mode, no signal, etc.)
     */
    object WaitingForNetwork : VpnWaitReason
}

@Parcelize
sealed interface VpnDisconnectError : Parcelable {

    /**
     * System failure establishing VPN connection involving multiple user profiles (or Dual Messenger).
     * On some Android devices setting up split tunneling can cause this error in multi-user scenarios.
     */
    data object InteractAcrossUsers: VpnDisconnectError

    /**
     * Android's VpnService error. See message for details.
     */
    data class ServiceError(val message: String): VpnDisconnectError

    /**
     * TUN interface couldn't be established or fails when reading/writing packets.
     */
    data class TunInterfaceError(val message: String?): VpnDisconnectError

    /**
     * VPN permission was not granted by the user. Make sure to request VPN permission before
     * attempting to connect.
     */
    data object VpnPermissionMissing: VpnDisconnectError
}

@Parcelize
data class PeerConnection(
    val protocol: VpnProtocol,
    val id: String,
    val entryAddr: InetSocketAddress,
): Parcelable